File: testing/Lab2/src/main/scala/function/TrigEvaluator.scala
package function

trait TTrigEvaluator {
  def sin(x: Double, eps: Double = 1E-6): Double
  def cos(x: Double, eps: Double = 1E-6): Double
  def csc(x: Double, eps: Double = 1E-6): Double
}

class TrigEvaluator(val sin: TSinus) extends TTrigEvaluator {
  def sin(x: Double, eps: Double = 1E-6): Double = sin.compute(x, eps)

  override def cos(x: Double, eps: Double = 1E-6): Double = if (x.isNaN) Double.NaN else x match {
    case Double.PositiveInfinity => Double.NaN
    case Double.NegativeInfinity => Double.NaN
    case _ => sin.compute(x + Math.PI / 2, eps)
  }

  override def csc(x: Double, eps: Double = 1E-6): Double = sin.compute(x, eps) match {
    case 0 => Double.NaN
    case a => 1 / a
  }
}

File: testing/Lab2/src/main/kotlin/function/TrigEvaluator.kt
package function

interface ITrigEvaluator {
    fun sin(x: Double, eps: Double = 1E-5): Double
    fun cos(x: Double, eps: Double = 1E-5): Double
    fun csc(x: Double, eps: Double = 1E-5): Double
}

class TrigEvaluator(private val sine: ISinus): ITrigEvaluator {
    override fun sin(x: Double, eps: Double): Double = sine.compute(x, eps)

    override fun cos(x: Double, eps: Double): Double =
        if (x.isNaN() || x.isInfinite() || eps.isNaN() || eps.isInfinite()) Double.NaN
        else sine.compute(x + Math.PI / 2, eps)

    override fun csc(x: Double, eps: Double): Double {
        val s = sine.compute(x, eps / 100)
        if (s == 0.0 || s.isNaN() || s.isInfinite())
            return Double.NaN
        return 1 / s
    }
}